File: src/com/cooksys/ssi/model/schema/Report.scala
package com.cooksys.ssi.model.schema

import java.sql.Date
import java.util.Base64

case class Report(title: String, data: String)

object Report {

  case class LayoutDrawing(jobId: Int)
  case class DetailDrawing(jobId: Int)
  case class ComputerDrawing(jobId: Int)
  case class Zone(jobId: Int)
  case class MaterialShipper(jobId: Int)
  case class ShipVia(jobId: Int)
  case class Shipment(jobId: Int)
  case class ManagementReview(start: Date, end: Date)
  case class ProductionSchedule(weekEnding: Date)
  case class SpecialtyItemsByJob(jobId: Int)
  case class SpecialtyItemsByPartType(partType: String)
  case class ShippingGroupShipper(jobId: Int, shippingGroupId: Int)
  case class JobShipment(shipmentId: Int)

  object Implicits extends Implicits

  trait Implicits extends JsonProtocol {

    implicit def byteArrayToBase64String(data: Array[Byte]): String =
      Base64.getMimeEncoder.encodeToString(data)

  }

  object JsonProtocol extends JsonProtocol

  trait JsonProtocol extends RootJsonProtocol {

    implicit val `JSON Report` = jsonFormat2(Report.apply)

    implicit val `JSON LayoutDrawing` = jsonFormat1(LayoutDrawing)
    implicit val `JSON DetailDrawing` = jsonFormat1(DetailDrawing)
    implicit val `JSON ComputerDrawing` = jsonFormat1(ComputerDrawing)
    implicit val `JSON Zone` = jsonFormat1(Zone)
    implicit val `JSON MaterialShipper` = jsonFormat1(MaterialShipper)
    implicit val `JSON ShipVia` = jsonFormat1(ShipVia)
    implicit val `JSON Shipment` = jsonFormat1(Shipment)
    implicit val `JSON ManagementReview` = jsonFormat2(ManagementReview)
    implicit val `JSON ProductionSchedule` = jsonFormat1(ProductionSchedule)
    implicit val `JSON SpecialtyItemsByJob` = jsonFormat1(SpecialtyItemsByJob)
    implicit val `JSON SpecialtyItemsByPartType` = jsonFormat1(SpecialtyItemsByPartType)
    implicit val `JSON ShippingGroupShipper` = jsonFormat2(ShippingGroupShipper)
    implicit val `JSON JobShipment` = jsonFormat1(JobShipment)

  }

}

File: ui/app/lib/services/api/Report.js
import { ApiService } from 'utils'
import { pdfConverter } from 'utils'

export default class Report extends ApiService {
  /*@ngInject*/
  constructor ($resource, endpoint, $q, $unpack) {
    super()

    var self = this;

    self.endpoint = $resource(endpoint + '/reports/:reportTitle', {}, {
        generate: { method: 'POST' }
      })

    self.productionSchedule = weekEnding =>
          self.endpoint
            .generate({ reportTitle: 'production-schedule' }, { weekEnding })
            .$promise
            .then(report => report.data).catch('some exception')

    self.managementReview = (start, end) =>
          self.endpoint
            .generate({ reportTitle: 'management-review' }, { start, end })
            .$promise
            .then(report => report.data).catch('some exception')

    self.rms = (jobId, shippingGroupId) =>
          self.endpoint
            .generate({ reportTitle: 'rms' }, { jobId, shippingGroupId })
            .$promise
            .then(report => report.data).catch('some exception')

    self.shipment = (shipmentId) =>
          self.endpoint
            .generate({ reportTitle: 'shipmentId' }, { shipmentId })
            .$promise
            .then(report => report.data).catch('some exception')

    self.jobShipments = (jobId) =>
          self.endpoint
            .generate({ reportTitle: 'job-shipments' }, { jobId })
            .$promise
            .then(report => report.data).catch('some exception')

    self.shipVia = (jobId) =>
          self.endpoint
            .generate({ reportTitle: 'ship-via' }, { jobId })
            .$promise
            .then(report => report.data).catch('some exception')

    self.materialShipper = (jobId) =>
          self.endpoint
            .generate({ reportTitle: 'material-shipper' }, { jobId })
            .$promise
            .then(report => report.data).catch('some exception')

    self.specialtyItemsByJob = (jobId) =>
          self.endpoint
            .generate({ reportTitle: 'specialty-items-by-job' }, { jobId })
            .$promise
            .then(report => report.data).catch('some exception')

    self.specialtyItemsByPartType = (partType) =>
          self.endpoint
            .generate({ reportTitle: 'specialty-items-by-part-type' }, { partType })
            .$promise
            .then(report => report.data).catch('some exception')

    self.layoutDrawing = (jobId) =>
      self.endpoint
            .generate({ reportTitle: 'layout-drawing' }, { jobId })
            .$promise
            .then(report => report.data).catch('some exception')

    self.detailDrawing = (jobId) =>
          self.endpoint
            .generate({ reportTitle: 'detail-drawing' }, { jobId })
            .$promise
            .then(report => report.data).catch('some exception')

    self.computerDrawing = (jobId) =>
          self.endpoint
            .generate({ reportTitle: 'computer-drawing' }, { jobId })
            .$promise
            .then(report => report.data).catch('some exception')

    self.zone = (jobId) =>
          self.endpoint
            .generate({ reportTitle: 'zone' }, { jobId })
            .$promise
            .then(report => report.data).catch('some exception')

  }

}
